### ISSUE TITLE
Model choice & dataset creation

### ISSUE THREAD

**Issue by StuartIanNaylor** [aegjoyce/microwakeword-custom-words#1]
I wondered if you shared your fixes as I am actually dubious the model can be fixed as that type of rolling window spectrogram is often just a toy example often provided as an introduction to wakeword.

I think its really pointless to use the model provided when SoTa wakeword models are available and opensource.
https://github.com/Qualcomm-AI-research/bcresnet the training script has been hardcoded for Google speech command datasets v1 and v2 but its very easy to hack that out and use any custom file based dataset in a similar file structure to https://github.com/google-research/google-research/tree/master/kws_streaming#training-on-custom-data

Its confusing as originally

> The streaming model performs inferences every 20 ms on the newest audio stride. The model is based on an [inceptional neural network](https://towardsdatascience.com/a-simple-guide-to-the-versions-of-the-inception-network-7fc52b863202?gi=6bc760f44aef) converted for streaming. Streaming and training the model uses heavily modified open-sourced code from [Google Research](https://github.com/google-research/google-research/tree/master/kws_streaming) found in the paper [Streaming Keyword Spotting on Mobile Devices](https://arxiv.org/pdf/2005.06720.pdf) by Rykabov, Kononenko, Subrahmanya, Visontai, and Laurenzo.

Is changed to

> The streaming model performs inferences every 30 ms, where the initial convolution layer strides over three 10 ms slices of audio. The model is a neural network using MixConv mixed depthwise convolutions suitable for streaming. Streaming and training the model uses heavily modified open-sourced code from Google Research found in the paper Streaming Keyword Spotting on Mobile Devices by Rykabov, Kononenko, Subrahmanya, Visontai, and Laurenzo.

As well as confusing the choice of model is surprising as https://arxiv.org/pdf/1907.09595 is a model by google that the authors excluded from Streaming Aware neural network models due to very high number of parameters and low accuracy results for what even with the MixNet-S it would seem to be, its just not that good of a wakeword model. I presume they ran into problems with the limited ML layers supported on esp32 via TF and converting https://github.com/google-research/google-research/blob/master/kws_streaming/models/inception.py to streaming as the google-research not managed to and left as 'not converted', which is a shame as it has 30k parameters as opposed to the 4.1m params of the arvix quote of MixNet-S, even if inputs have been adjested for spectrogram. Even then quantised spectrogram, MFCC has been choice to reduce input size whilst not losing features, for decades and near nowhere else apart from toy training scripts will use heavy load spectrogram over MFCC.

So when its a model the likes of Rykabov, Kononenko, Subrahmanya, Visontai, and Laurenzo don't even bother to test as a streaming model, its likely it might not be a good choice, but also if you are going to use esp32 the Onnx ML implementation garners far more Espressif dev attention providing far more ML operators/layers https://github.com/espressif/esp-dl/blob/master/operator_support_state.md than the rather static https://github.com/espressif/esp-nn of far fewer than seem to have hard coded and fixed inputs.

So to stop wasting any further time on what might be a bad choice of model and possible platform to dispel "Training a model that works well is still very difficult, as it typically requires experimentation with hyper-parameters and sample generation settings. Please share any insights you find for training a good model!"

As from my own previous experimentation with training ds-cnn, crnn, cnn and what is now my preference due to the lowest number of parameters for also highest accuracy results is a bcresnet.
None of them needed experimentation with bcresnet and not sure what sample generation settings setting mean but any TTS that can provide good quality 16khz speech samples with a range of prosody, needs no more experimentation.

When it comes to classification models the accuracy of your model is highly dependent on the quality of the model and for some reason this myth that an almost binary classification of positive & negative classification with an additional 'noise' class makes a good dataset.
It has nothing to do with playing hyper-parameters as a single training run will give an overfitting training curve due to 2 hugely problematic methods employed.

1... The piper sample generator provides only 1000 samples 500 of each gender and apart from distinguishable gender change there is very little prosody variation than the same american english voice. As a rule of thumb an approx dataset size should be several orders of magnitude larger than the models parameters.

2... Huge class imbalance with a positive class that contains the features of a single wakeword and the features of the rest of a full language in a single negative classification. You can test this out by simply running tests on the Google speech command datasets v1 and v2 as the introduction of more classes and few words in the negative classification, gives a very different more gradual training curve than the hockey stick of a relatively binary classes of just positive vs negative. Also the cross entropy of the classes is much lower with just 2 classes of a wakeword and all the other sounding words in a language. This also contributes to a hockey stick training curve and training accuracies of near 100% as the wakeword to non wakeword classes is so dissimilar in features that its akin to throwing a tennis ball into two swimming pools at 1m as training accuracy is relative to class cross-entropy. Add more cross-entropy as you will reduce the target size to a dart and dartboard bullseye at 1m where you might hit only 75% but which one is more accurate as training accuracy is purely relative.
If like Google speech command datasets that do have more cross entropy they will complete train with less accuracy but in operation be more accurate as the cross entropy creates a smaller target and forces the model to find further features to differentiate class.
It deliberately includes words such as 'No' and 'Go' that create very similar spectrograms as its a benchmark test dataset.

I am not sure why the Pipe sample generator creates such low prosody variation as the model https://k2-fsa.github.io/sherpa/onnx/tts/pretrained_models/vits.html#vits-piper-en-us-libritts-r-medium-english-904-speakers is piper but I guess bigger maybe with a stronger vocoder but the latter gives far more prosody variation and noticeable better quality.
No single TTS gives a prosody variation of a truly global selection but using https://github.com/idiap/coqui-ai-TTS xVitts to force english through other languages will garner accents and its a great tool.
I also use piper from above and whilst in sherpa the https://k2-fsa.github.io/sherpa/onnx/tts/pretrained_models/kokoro.html#kokoro-multi-lang-v1-1-chinese-english-103-speakers https://k2-fsa.github.io/sherpa/onnx/tts/pretrained_models/vits.html#vctk-english-multi-speaker-109-speakers  & https://k2-fsa.github.io/sherpa/onnx/tts/pretrained_models/kokoro.html#kokoro-multi-lang-v1-0-chinese-english-53-speakers have different prosody if with less speakers.
https://github.com/netease-youdao/EmotiVoice/wiki/%F0%9F%98%8A-voice-wiki-page is good with nearly 2000 voices if english accent only.

Then you have what you should use as negative dataset and how to get all those voices say multiple unique features so the model is focusing on features and not voice variation.
https://drive.google.com/file/d/1zpECiZ8G9ZQeWn1WI-oQAAgVf6eNpBAV/view?usp=drive_link is a text file in the following format

```
word,scount,p0,p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,p13,p14,p15,p16,p17,p18,p19,p20,p21,p22,p23,p24,p25,p26,p27,p28,p29,pcount
aaberg,2,AA1,B,ER0,G,,,,,,,,,,,,,,,,,,,,,,,,,,,4
aachen,2,AA1,K,AH0,N,,,,,,,,,,,,,,,,,,,,,,,,,,,4
aachener,3,AA1,K,AH0,N,ER0,,,,,,,,,,,,,,,,,,,,,,,,,,5
aaker,2,AA1,K,ER0,,,,,,,,,,,,,,,,,,,,,,,,,,,,3
aalborg,2,AA1,L,B,AO2,R,G,,,,,,,,,,,,,,,,,,,,,,,,,6
aalburg,2,AA1,L,B,ER0,G,,,,,,,,,,,,,,,,,,,,,,,,,,5
aalen,2,AA1,L,AH0,N,,,,,,,,,,,,,,,,,,,,,,,,,,,4
```

Language is an emphasis on syllables and starting phones of a syllable are always the strongest and will be the strongest representation in a spectrogram. Realtime tools to find syllables I have found to be pretty weak, where https://github.com/repp/big-phoney seems to be the best but very slow so I have already run through the whole cmu dictionary so you can just import into a Sqlite database to run syllable and phone queries (DB Browser for SQLite is a great tool for this).

Likely linguists would be able to pick a subset, but I am no linguist but a brute force query `select "word" from "cmu-words_meta" where "scount" = 3 group by "p0","p1","p2","p3","p4"` actually returns a  22,273 row dataset that with all the voices you have can easily make a good dataset size.
I am just going to stick with a basic binary like classification of positive vs negative even with a class imbalance as if you do use another model that what has been proposed like me will will likely find your results to be more accurate.
I say that because what happens next in the training script in my mind is pure weird as they apply a dataset of RiRs all recorded at 1.5m that includes many locations such as forrests and shopping malls for a smart-speaker to be operated at any distance up to 3m in what will likely be average domestic room sizes.
I have a hacky script that uses GpuRiR and various augmentation and mixes in noise to a small level as above approx 30% it makes little difference but starts to have huge effect on accuracy so 30% or less noise is optimal as you can train in a a bit of resilience but its much better to have pre noise removal.
 https://github.com/StuartIanNaylor/wake_word_capture/blob/main/augment/augment.py is rough and ready but will show how to apply GpuRir to create Rirs that are likely in the likely room sizes we will be using.

I have a currated noise dataset I use https://drive.google.com/file/d/1tY6qkLSTz3cdOnYRuBxwIM5vj-w4yTuH/view?usp=drive_link
If using cloning TTS I already have a download collection of accents from the excellent https://accent.gmu.edu/
https://drive.google.com/file/d/1gPiz--326JJSO_1CspNt97VPW1G95jlW/view?usp=drive_link

But create your own dataset on your own model as likely you will find it massively more accurate. I agree that "I think this is underselling it!" and will be even stronger as from many responses that make little sense or seem to have any truth, its pretty obvious code has been hoovered and refactored and rebranded as own with little dev and understanding of some fairly basic Ml.
The model and dataset creation are Fubar and its probably impossible to do any worse, so really do suggest pick a model and create your own dataset.

If you are going to get more geeky then you are going to deliberate add more cross entropy and I do this by choosing phonetically similar words of the same syllable length and add 2 more 'like' classes for the wakeword of like sounding start syllable and end syllable as two classes.
Also to balance that as my wakewords are usually 3 syllable at least by choice I add further negative of 1syl & 2syl words in one class and 4syl and above in a 3rd negative class with a noise class from the noise dataset provided.
Each additional class adds very few parameters so models have negligible size or compute gains so its very possible to add further unique wakeword and accompanying classes.

You might also want to use your wakeword dataset to add further data to the https://github.com/breizhn/DTLN dataset and train in your wakeword or even your own voice for further accuracy.
DTLN is a great low compute speech enhancment filter that will run with a wake word model even on a Raspberry Pi Zero2 and has an implementation ready to go in https://github.com/SaneBow/PiDTLN
